<a href="https://colab.research.google.com/github/dougscohen/DS-Unit-1-Sprint-2-Statistics/blob/master/module1/LS_DS_121_Statistics_Probability_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img align="left" src="https://lever-client-logos.s3.amazonaws.com/864372b1-534c-480e-acd5-9711f850815c-1524247202159.png" width=200>
<br></br>
<br></br>

## *Data Science Unit 1 Sprint 2 Assignment 1*

# Apply the t-test to real data

Your assignment is to determine which issues have "statistically significant" differences between political parties in this [1980s congressional voting data](https://archive.ics.uci.edu/ml/datasets/Congressional+Voting+Records). The data consists of 435 instances (one for each congressperson), a class (democrat or republican), and 16 binary attributes (yes or no for voting for or against certain issues). Be aware - there are missing values!

Your goals:

1. Load and clean the data (or determine the best method to drop observations when running tests)
2. Using hypothesis testing, find an issue that democrats support more than republicans with p < 0.01
3. Using hypothesis testing, find an issue that republicans support more than democrats with p < 0.01
4. Using hypothesis testing, find an issue where the difference between republicans and democrats has p > 0.1 (i.e. there may not be much of a difference)

Note that this data will involve *2 sample* t-tests, because you're comparing averages across two groups (republicans and democrats) rather than a single group against a null hypothesis.

Stretch goals:

1. Refactor your code into functions so it's easy to rerun with arbitrary variables
2. Apply hypothesis testing to your personal project data (for the purposes of this notebook you can type a summary of the hypothesis you formed and tested)

In [0]:
# Import Pandas

import pandas as pd

# load in the dataset

column_headers = ['Class Name','handicapped-infants', 
                  'water-project-cost-sharing', 
                  'adoption-of-the-budget-resolution', 'physician-fee-freeze', 
                  'el-salvador-aid', 'religious-groups-in-schools', 
                  'anti-satellite-test-ban', 'aid-to-nicaraguan-contras', 
                  'mx-missile', 'immigration', 'synfuels-corporation-cutback', 
                  'education-spending' , 'superfund-right-to-sue', 'crime', 
                  'duty-free-exports', 'export-administration-act-south-afric']

df = pd.read_csv('https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data', names=column_headers)

df.head()

Class Name  ... export-administration-act-south-afric
0  republican  ...                                     y
1  republican  ...                                     ?
2    democrat  ...                                     n
3    democrat  ...                                     y
4    democrat  ...                                     y

[5 rows x 17 columns]

In [0]:
# Ensure there are 435 rows with 16 binary attributes 
# (first column isn't binary)

df.shape

(435, 17)

In [0]:
# Change values to numbers

import numpy as np

df = df.replace({'y': 1, 'n': 0, '?': np.NaN})

df.head()

Class Name  ...  export-administration-act-south-afric
0  republican  ...                                    1.0
1  republican  ...                                    NaN
2    democrat  ...                                    0.0
3    democrat  ...                                    1.0
4    democrat  ...                                    1.0

[5 rows x 17 columns]

In [0]:
# Create a subset of the dataset that is just Republicans

rep = df[df['Class Name'] == 'republican']

rep.head()

Class Name  ...  export-administration-act-south-afric
0   republican  ...                                    1.0
1   republican  ...                                    NaN
7   republican  ...                                    1.0
8   republican  ...                                    1.0
10  republican  ...                                    0.0

[5 rows x 17 columns]

In [0]:
# Create a subset of the dataset that is just Democrats

dem = df[df['Class Name'] == 'democrat']

dem.head()

Class Name  ...  export-administration-act-south-afric
2   democrat  ...                                    0.0
3   democrat  ...                                    1.0
4   democrat  ...                                    1.0
5   democrat  ...                                    1.0
6   democrat  ...                                    1.0

[5 rows x 17 columns]

In [0]:
df['Class Name'].value_counts()

democrat      267
republican    168
Name: Class Name, dtype: int64

In [0]:
dem.shape

(267, 17)

In [0]:
rep.shape

(168, 17)

## Find an issue that democrats support more than republicans with p < 0.01

1) Null hypothesis: Democratic support for the "duty-free-exports" bill is the same as the Republican support for that bill

2) Alternative Hypothesis: Democratic support for the "duty-free-exports" bill is the NOT the same as the Republican support for that bill

3) Confidence: 99%

In [0]:
from scipy import stats

stats.ttest_ind(dem['duty-free-exports'], rep['duty-free-exports'], nan_policy='omit')

Ttest_indResult(statistic=12.853146132542978, pvalue=5.997697174347365e-32)

4) Conclusion: Based on a t-statistic of ~12.85 and a p-value of ~0, I reject the null hypothesis that democrats support the "duty-free-exports" bill at the same level. And I suggest the alternative that democratic and republican levels of support for this bill are different.

## Find an issue that republicans support more than democrats with p < 0.01

1) Null hypothesis: Democratic support for the "physician-fee-freeze" bill is the same as the Republican support for that bill

2) Alternative Hypothesis: Democratic support for the "physician-fee-freeze" bill is the NOT the same as the Republican support for that bill

3) Confidence: 99%

In [0]:
from scipy import stats

stats.ttest_ind(rep['physician-fee-freeze'], dem['physician-fee-freeze'], nan_policy='omit')

Ttest_indResult(statistic=49.36708157301406, pvalue=1.994262314074344e-177)

4) Conclusion: Based on a t-statistic of ~49.37 and a p-value of ~0, I reject the null hypothesis that democrats support the "physician-fee-freeze" bill at the same level. And I suggest the alternative that democratic and republican levels of support for this bill are different.

## Find an issue where the difference between republicans and democrats has p > 0.1 (i.e. there may not be much of a difference)

1) Null hypothesis: Democratic support for the "water-project-cost-sharing" bill is the same as the Republican support for that bill

2) Alternative Hypothesis: Democratic support for the "water-project-cost-sharing" bill is the NOT the same as the Republican support for that bill

3) Confidence: 99%

In [0]:
from scipy import stats

stats.ttest_ind(dem['water-project-cost-sharing'], rep['water-project-cost-sharing'], nan_policy='omit')

Ttest_indResult(statistic=-0.08896538137868286, pvalue=0.9291556823993485)

4) Conclusion: Based on a t-statistic of -0.08897 and a p-value of 0.92916, I fail to reject the null hypothesis that democrats support the "water-project-cost-sharing" bill at the same level.

# Stretch goals:

1) Refactor your code into functions so it's easy to rerun with arbitrary variables

2) Apply hypothesis testing to your personal project data (for the purposes of this notebook you can type a summary of the hypothesis you formed and tested)

In [0]:
# Make a function that does the same thing as the code above, but more easily.

def quick_ttest(var_a, var_b):
  x = stats.ttest_ind(dem[var_a], rep[var_b], nan_policy='omit')
  return x

In [0]:
quick_ttest('water-project-cost-sharing', 'physician-fee-freeze')

Ttest_indResult(statistic=-12.248400713110872, pvalue=1.5689751787467094e-29)

Null hypothesis: Democratic support for the "crime" bill occurs at the same rate as the Republican support for the "religos-groups-in-school" bill.

Alternative Hypothesis: Democratic support for the "crime" bill does NOT occur at same rate as the Republican support for the "religous-groups-in-school" bill.

Confidence: 95%

In [0]:
quick_ttest('crime', 'religious-groups-in-schools')

Ttest_indResult(statistic=-13.135205640116128, pvalue=2.860855669046361e-33)

Conclusion: Based on a t-statistic of -13.135 and a p-value of ~0, I reject the null hypothesis that democratic support of the "crime" bill is the same as the republican support of the "religoius-groups-in-school" bill. And I suggest the alternative that the level at which democrats support the "crime" bill is different than the level at which republicans support the "religious-groups-in-school" bill.